openedx/features/effort_estimation/block_transformers.py

```diff
@@ -0,0 +1,225 @@
+        # record this duration separately, to use instead of the whole video duration if the client supports it.
+        clip_duration = (xblock.end_time - xblock.start_time).total_seconds()
+        if clip_duration > 0:
+            block_structure.set_transformer_block_field(block_key, cls, cls.VIDEO_CLIP_DURATION, clip_duration)
+
+    def transform(self, usage_info, block_structure):
+        # Skip any transformation if our collection phase said to
+        cls = EffortEstimationTransformer
+        if block_structure.get_transformer_data(cls, cls.DISABLE_ESTIMATION, default=False):
+            return
+
+        # Get bucket for this experiment. 0 is no estimate. 1 is only on sections. 2 is only on subsections.
+        # If we drop the experiment but keep the feature - we should leave estimates on all blocks, and just have the
+        # UI show the estimate at the level they care about.
+        bucket = EFFORT_ESTIMATION_LOCATION_FLAG.get_bucket(course_key=usage_info.course_key)
+        if bucket == 0:
+            return  # early exit, we don't need to do any of this work
+
+        # We're good to continue and make user-specific estimates based on collected data
+        for block_key in block_structure.post_order_traversal():
+            category = block_structure.get_xblock_field(block_key, 'category')
+
+            # These estimation methods should return a tuple of (a number in seconds, an activity count)
+            estimations = {
+                'chapter': self._estimate_children_effort,
+                'course': self._estimate_children_effort,
+                'html': self._estimate_html_effort,
+                'sequential': self._estimate_children_effort,
+                'vertical': self._estimate_vertical_effort,
+                'video': self._estimate_video_effort,
+            }
+
+            if category not in estimations:
+                continue
+
+            time, activities = estimations[category](usage_info, block_structure, block_key)
+
+            if time is not None:
+                # We take the ceiling of the estimate here just for cleanliness. Losing the fractional seconds does
+                # technically make our estimate less accurate, especially as we combine these values in parents.
+                # But easier to present a simple integer to any consumers, and precise to-the-second accuracy on our
+                # estimate is not a primary goal.
+                block_structure.override_xblock_field(block_key, self.EFFORT_TIME, math.ceil(time))
+
+            if activities is not None:
+                block_structure.override_xblock_field(block_key, self.EFFORT_ACTIVITIES, activities)
+
+        # Second pass to clear out collected estimate on levels we don't want to share. Just an easy way to test
+        # estimates at different levels, per the experiment. If we make this real code and not an experiment, just
+        # drop this whole section.
```

Comment:
should there be a TODO of some sort here?  
or perhaps we make a ticket of things to address if we make this code real and put the ticket number here?

Reply:
I made https://openedx.atlassian.net/browse/AA-659 to hold that stuff and added a couple references to it around where I thought it might help.